cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+
+		switch (zio_cmd->hdr.opcode) {
+			case ZVOL_OPCODE_HANDSHAKE:
+			case ZVOL_OPCODE_WRITE:
+			case ZVOL_OPCODE_SYNC:
+				zinfo->write_req_ack_cnt++;
+				/* Send handsake ack */
+				break;
+			case ZVOL_OPCODE_READ:
+				printf("ACK for op:%d with seq-id %ld\n",
+				    zio_cmd->hdr.opcode, zio_cmd->hdr.io_seq);
+				/* Send data read from disk */
+				rc = uzfs_zvol_socket_write(zio_cmd->conn,
+				    zio_cmd->buf,
+				    (sizeof (char) * zio_cmd->hdr.len));
+				if (rc == -1) {
+					ZREPL_ERRLOG("socket write err :%d\n",
+					    errno);
+					ASSERT(0);
+					goto exit;
+				}
+				zinfo->read_req_ack_cnt++;
+				break;
+
+			default:
+				VERIFY(!"Should be a valid opcode");
+				break;
+		}
+		zio_cmd_free(&zio_cmd, zio_cmd->hdr.opcode);
+	}
+exit:
+	close(fd);
+	while (!STAILQ_EMPTY(&zinfo->complete_queue)) {
+		zio_cmd = STAILQ_FIRST(&zinfo->complete_queue);
+		STAILQ_REMOVE_HEAD(&zinfo->complete_queue, cmd_link);
+		zio_cmd_free(&zio_cmd, zio_cmd->hdr.opcode);
+	}
+	uzfs_zinfo_drop_refcnt(zinfo, false);
+
+	printf("uzfs_zvol_io_ack_sender thread exiting\n");
+	zk_thread_exit();
+}
+
+static void
+uzfs_zrepl_open_log(void)
+{
+	openlog("zrepl", LOG_PID, LOG_LOCAL7);
+}
+
+static void
+uzfs_zrepl_close_log(void)
+{
+	closelog();
+}
+
+static void
+uzfs_zrepl_walk_pool_directory(void)
+{
+	spa_t	*spa = NULL;
+	spa_t	*sp = NULL;
+	int	rc = 0;
+
+	mutex_enter(&spa_namespace_lock);
+	while ((spa = spa_next(spa)) != NULL) {
+		(void) printf(">>>>>>>\t%s\n", spa_name(spa));
+		rc = spa_open(spa_name(spa), &sp, spa);
+		if (rc == 0) {
+			spa_close(spa, spa);
+		}
+	}
+	mutex_exit(&spa_namespace_lock);
+}
+
+/*
+ * Main function for replica.
+ */
+int
+main(void)
+{
+	int 		rc;
+	kthread_t 	*conn_accpt_thrd;
+	kthread_t 	*uzfs_mgmt_thread;
+
+
+	pthread_t slf = pthread_self();
+	snprintf(tinfo, sizeof (tinfo), "m#%d.%d",
+	    (int)(((uint64_t *)slf)[0]), getpid());
+
+	zfs_arc_max = (512 << 20);
+	zfs_arc_min = (256 << 20);
+	printf("zarcmax: %lu zarcmin:%lu\n", zfs_arc_max, zfs_arc_min);
+
+	rc = uzfs_init();
+	uzfs_zrepl_open_log();
+	if (rc != 0) {
+		printf("initialization errored.. %d\n", rc);
+		return (-1);
+	}
+
+	rc = pthread_mutex_init(&zvol_list_mutex, NULL);
+	if (rc != 0) {
+		ZREPL_ERRLOG("zvol_global mutex_init() failed\n");
+		return (-1);
+	}
+
+	uzfs_zrepl_walk_pool_directory();
+	sleep(5);
+
+	if (libuzfs_ioctl_init() < 0) {
+		ZREPL_ERRLOG("Failed to initialize libuzfs ioctl\n");
+		(void) fprintf(stderr, "%s",
+		    "failed to initialize libuzfs ioctl\n");
+		goto initialize_error;
+	}
+
+	conn_accpt_thrd = zk_thread_create(NULL, 0,
+	    (thread_func_t)uzfs_zvol_io_conn_acceptor,
+	    NULL, 0, NULL, TS_RUN, 0,
+	    PTHREAD_CREATE_DETACHED);
+	VERIFY3P(conn_accpt_thrd, !=, NULL);
+
+	uzfs_mgmt_thread = zk_thread_create(NULL, 0,
+	    (thread_func_t)uzfs_zvol_mgmt_thread,
+	    NULL, 0, NULL, TS_RUN, 0,
+	    PTHREAD_CREATE_DETACHED);
+	VERIFY3P(uzfs_mgmt_thread, !=, NULL);
+
+	while (1) {
+		sleep(5);
+	}
```

Comment:
I would expect the main thread to do at least some signal management like:
```
while (notexit) {
    sleep(...)
}
```
and before that set signal mask and handlers so that when someone press ctrl+c the whole process is cleanly shutdown, finishing all work which needs to be finished etc.

Reply:
I agree with you. Focus was mainly on connections and IOs. These things like memory limit, signal handling etc ...left as it is. But these are equally important. Will fix these issues via separate bug.

Reply:
👍

Reply:
Jan, 

Replica will run as a background process, having said that mean we do not need to handle CTRL + C. Me and Vitta discussed here and found that we need to handle SIGPIPE signal, all other signals will be taken care in terms of replica failure, which will cause all threads and memory to go for toss. I hope this is inline with what behaviour we want out of replica during shutdown/crash.